blueocean-bitbucket-pipeline/src/main/java/io/jenkins/blueocean/blueocean_bitbucket_pipeline/AbstractBitbucketScm.java

```diff
@@ -0,0 +1,276 @@
+            } else {
+                CredentialsUtils.updateCredentialsInUserStore(
+                        bbCredentials, credential, authenticatedUser, getDomainId(),
+                        ImmutableList.<DomainSpecification>of(new BlueOceanDomainSpecification()));
+            }
+
+            return createResponse(credential.getId());
+        }catch (IOException e){
+            throw new ServiceException.UnexpectedErrorException(e.getMessage());
+        }
+    }
+
+    private void validate(String userName, String password, String apiUrl){
+        List<ErrorMessage.Error> errorList = new ArrayList<>();
+        if(StringUtils.isBlank(userName)){
+            errorList.add(new ErrorMessage.Error("userName",
+                    ErrorMessage.Error.ErrorCodes.MISSING.toString(),
+                    "userName is required parameter"));
+        }
+        if(StringUtils.isBlank(password)){
+            errorList.add(new ErrorMessage.Error("password",
+                    ErrorMessage.Error.ErrorCodes.MISSING.toString(),
+                    "password is required parameter"));
+        }
+        if(StringUtils.isBlank(apiUrl)){
+            errorList.add(new ErrorMessage.Error("apiUrl",
+                    ErrorMessage.Error.ErrorCodes.MISSING.toString(),
+                    "apiUrl is required parameter"));
+        }else{
+            try {
+                new URL(apiUrl);
+            } catch (MalformedURLException e) {
+                errorList.add(new ErrorMessage.Error("apiUrl",
+                        ErrorMessage.Error.ErrorCodes.INVALID.toString(),
+                        "apiUrl must be a URL: "+e.getMessage()));
+            }
+        }
+        if(!errorList.isEmpty()){
+            throw new ServiceException.BadRequestException(
+                    new ErrorMessage(400, "Invalid request").addAll(errorList));
+        }
+    }
+
+    public static BitbucketApi getApi(String apiUrl, StandardUsernamePasswordCredentials credentials){
+        BitbucketApiFactory apiFactory = BitbucketApiFactory.resolve(apiUrl);
+        if(apiFactory == null){
+            throw new ServiceException.UnexpectedErrorException("BitbucketApiFactory to handle apiUrl "+apiUrl+" not found");
+        }
+
+        return apiFactory.newInstance(apiUrl, credentials);
+    }
+
+    protected  abstract @Nonnull String createCredentialId(@Nonnull String apiUrl);
+
+    protected abstract @Nonnull String getDomainId();
+
+    protected StaplerRequest getStaplerRequest(){
+        StaplerRequest request = Stapler.getCurrentRequest();
+        Preconditions.checkNotNull(request, "Must be called in HTTP request context");
+        return request;
+    }
+
+    protected String getApiUrlParameter(){
+        return getApiUrlParameter(getStaplerRequest());
+    }
+
+    private String getApiUrlParameter(StaplerRequest request){
+        return request.getParameter("apiUrl");
+    }
```

Comment:
In GHE we had some annoyances with the `apiUrl` containing the trailing slash, or not, which might bypass duplicate detection inappropriately. I think there is some logic downstream to detect this elsewhere in the code, although it might have a bug in it.  Would it be better to add a method here like `String getCanonicalApiUrlParameterValue()` which returns the `apiUrl` with trailing slash always added? Then all downstream logic wouldn't have to worry about it. To be extra safe, we could add the same logic in endpoint handling code in case someone adds an endpoint manually via classic?